### PR TITLE
[Validator] Propagate embedded groups for collection validator

### PR DESCRIPTION
**Pull request by blazarecki**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17675 |
| License | MIT |
| Doc PR | no |

This PR fix the issue describe in #17675

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/CollectionValidator.php [hunk 1]
```diff
@@ -60,7 +60,7 @@ public function validate($value, Constraint $constraint)
                         $context->getValidator()
                             ->inContext($context)
                             ->atPath('['.$field.']')
-                            ->validate($value[$field], $fieldConstraint->constraints);
+                            ->validate($value[$field], $fieldConstraint->constraints, $context->getGroup());
                     } else {
                         // 2.4 API
```
Comment (HeahDude) [inline src/Symfony/Component/Validator/Constraints/CollectionValidator.php]: It's not the scope of this PR but shouldn't this be remove in 2.7 ?

**Comment by blazarecki** [comment]
The issue is fixed for the current API. 
But it fail on the 2.4 API.

For the 2.4 API I've try to use the group from context and from the constraint and both don't work.

Any ideas ?

Here the phpunit errors:

```
There were 2 failures:

1) Symfony\Component\Validator\Tests\LegacyValidatorTest::testValidateMultipleGroupsForCollectionConstraint
Failed asserting that actual size 4 matches expected size 2.

/Users/benjamin.lazarecki/Github/symfony/src/Symfony/Component/Validator/Tests/Validator/AbstractValidatorTest.php:1108

2) Symfony\Component\Validator\Tests\LegacyValidatorTest::testValidateMultipleGroupsForAllConstraint
Failed asserting that actual size 4 matches expected size 2.

/Users/benjamin.lazarecki/Github/symfony/src/Symfony/Component/Validator/Tests/Validator/AbstractValidatorTest.php:1132
```

File: src/Symfony/Component/Validator/Constraints/CollectionValidator.php (hunk 1 above)
Comment (blazarecki) [inline src/Symfony/Component/Validator/Constraints/CollectionValidator.php]: Why removed ? I check on the 2.8 and 3.0 branch and the code is the same.

**Comment by backbone87** [comment]
There are also problem with GroupSequence.

```php
    public function testValidateGroupSequenceForCollectionConstraint()
    {
        $entity = new Entity();
        $entity->firstName = array('baz' => 2);

        $sequence = new GroupSequence(array('bar', 'foo'));
        $this->metadata->setGroupSequence($sequence);

        $this->metadata->addPropertyConstraint('firstName', new Collection(
            array(
                'fields' => array(
                    'baz' => array(
                        new Range(array('min' => 3, 'minMessage' => 'Group foo', 'groups' => 'foo')),
                        new Range(array('min' => 5, 'minMessage' => 'Group bar', 'groups' => 'bar')),
                    ),
                ),
            )
        ));

        $violations = $this->validate($entity, null, 'Default');

        $this->assertCount(1, $violations);
        $this->assertSame('Group bar', $violations[0]->getMessage());
    }
```

This test should fail, but I did not actually run it.

**Comment by blazarecki** [comment]
@backbone87 I add the test for group sequence.

**Comment by backbone87** [comment]
@blazarecki what should c1311b8 fix?

**Comment by blazarecki** [comment]
Hi @backbone87,

With this fix, now the current group is propagate to the embedded constraints. 
Then the `executeConstraintValidators` method, filters only the constraint that match the given group.

**Comment by dunglas** [comment]
:+1: 

`GroupSequence` can be fixed in another PR.

ping @symfony/deciders

**Comment by dunglas** [comment]
Status: reviewed

**Comment by fabpot** [comment]
:+1:

File: src/Symfony/Component/Validator/Constraints/AllValidator.php [hunk 2]
```diff
@@ -44,12 +44,12 @@ public function validate($value, Constraint $constraint)
             $validator = $context->getValidator()->inContext($context);
 
             foreach ($value as $key => $element) {
-                $validator->atPath('['.$key.']')->validate($element, $constraint->constraints);
+                $validator->atPath('['.$key.']')->validate($element, $constraint->constraints, $constraint->groups);
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/AllValidator.php]: Shouldn't this be `$context->getGroup()` too?

Comment (blazarecki) [inline src/Symfony/Component/Validator/Constraints/AllValidator.php]: Hi, @xabbuh 

I try to use the context but the test fail, it's the only way I find to make it work:

``` 
1) Symfony\Component\Validator\Tests\Validator\LegacyValidator2Dot5ApiTest::testValidateMultipleGroupsForAllConstraint
Failed asserting that actual size 1 matches expected size 2.

symfony/src/Symfony/Component/Validator/Tests/Validator/AbstractValidatorTest.php:1129

2) Symfony\Component\Validator\Tests\Validator\LegacyValidatorLegacyApiTest::testValidateMultipleGroupsForAllConstraint
Failed asserting that actual size 1 matches expected size 2.

symfony/src/Symfony/Component/Validator/Tests/Validator/AbstractValidatorTest.php:1129

3) Symfony\Component\Validator\Tests\Validator\RecursiveValidator2Dot5ApiTest::testValidateMultipleGroupsForAllConstraint
Failed asserting that actual size 1 matches expected size 2.
```

Do you have another idea ?

**Comment by blazarecki** [comment]
@webmozart what do you think of this PR ?

**Comment by blazarecki** [comment]
@javiereguiluz any new about this PR ? Is there something missing ? Thanks.

File: src/Symfony/Component/Validator/ExecutionContext.php [hunk 3]
```diff
@@ -271,7 +271,15 @@ public function getMetadataFactory()
      */
     private function executeConstraintValidators($value, array $constraints)
     {
+        // Filter constraints with the context group.
+        $filteredConstraints = array();
         foreach ($constraints as $constraint) {
+            if (in_array($this->getGroup(), $constraint->groups)) {
```
Comment (webmozart) [inline src/Symfony/Component/Validator/ExecutionContext.php]: This looks weird to me. Why is this necessary?

File: src/Symfony/Component/Validator/Tests/Validator/AbstractValidatorTest.php [hunk 4]
```diff
@@ -1081,6 +1084,50 @@ public function testValidateMultipleGroups()
         $this->assertCount(2, $violations);
     }
 
+    public function testValidateMultipleGroupsForCollectionConstraint()
```
Comment (webmozart) [inline src/Symfony/Component/Validator/Tests/Validator/AbstractValidatorTest.php]: This test belongs in CollectionValidatorTest. Also, I'm missing tests for what is happening when one or more groups are set on the `Collection` constraint itself. You should systematically test the different permutations:

* `Collection` group equals that of the constraints
* `Collection` group does not equal that of the constraints
* `Collection` groups are a superset of the groups of the constraints
* `Collection` groups are a subset of the groups of the constraints
* `Collection` groups and constraing groups don't overlap

The same applies to AllValidator.

**Comment by webmozart** [comment]
Thank you for working on this PR @blazarecki! I think that in general, this functionality is nice. This PR needs some work however to ensure that we don't introduce bugs with this functionality.

Status: Needs work

**Comment by blazarecki** [comment]
Thanks for your feedback, I'll work on it.

**Comment by fabpot** [comment]
@blazarecki What's the status of this PR? Can you answer @webmozart's question?

File: src/Symfony/Component/Validator/ExecutionContext.php (hunk 3 above)
Comment (blazarecki) [inline src/Symfony/Component/Validator/ExecutionContext.php]: It's necessary because for embedded constraint with multiple validation groups, the validator is execute for each validation groups.

With this example:

```
$entity = new Entity();
$entity->firstName = array('baz' => 2);

$this->metadata->addPropertyConstraint('firstName', new Collection(
    array(
        'fields' => array(
            'baz' => array(
                new Range(array('min' => 3, 'groups' => array('foo', 'bar'))),
            ),
        ),
    )
));

$violations = $this->validate($entity, null, array('foo', 'bar'));
```

We have two violations but we want only one. 

Is there another way to fix that ?

**Comment by blazarecki** [comment]
@webmozart what kind of work is needed for this PR to ensure that we don't introduce bugs with this functionality?

File: src/Symfony/Component/Validator/Tests/Validator/AbstractValidatorTest.php (hunk 4 above)
Comment (xabbuh) [inline src/Symfony/Component/Validator/Tests/Validator/AbstractValidatorTest.php]: @blazarecki Can you address this comment?

**Comment by nicolas-grekas** [comment]
@blazarecki rebase needed please. Can you please check again and see what's missing? Anyone else to help review?

**Comment by blazarecki** [comment]
Gitub cannot find the repository and branch so I create a new PR https://github.com/symfony/symfony/pull/25888
